spring-integration-core/src/main/java/org/springframework/integration/scattergather/ScatterGatherHandler.java

```diff
@@ -182,15 +178,27 @@ protected Object handleRequestMessage(Message<?> requestMessage) {
 				getMessageBuilderFactory()
 						.fromMessage(requestMessage)
 						.setHeader(GATHER_RESULT_CHANNEL, gatherResultChannel)
-						.setHeader(ORIGINAL_REPLY_CHANNEL, requestMessageHeaders.getReplyChannel())
-						.setHeader(ORIGINAL_ERROR_CHANNEL, requestMessageHeaders.getErrorChannel())
 						.setReplyChannel(this.gatherChannel)
 						.setErrorChannelName(this.errorChannelName)
 						.build();
 
 		this.messagingTemplate.send(this.scatterChannel, scatterMessage);
 
-		return gatherResultChannel.receive(this.gatherTimeout);
+		Message<?> gatherResult = gatherResultChannel.receive(this.gatherTimeout);
+
+		if (gatherResult != null) {
+			if (gatherResult instanceof ErrorMessage) {
+				throw (RuntimeException) gatherResult.getPayload();
+			}
+			return getMessageBuilderFactory()
+					.fromMessage(gatherResult)
+					.setHeader(GATHER_RESULT_CHANNEL, requestMessageHeaders.get(GATHER_RESULT_CHANNEL))
+					.setHeader(MessageHeaders.REPLY_CHANNEL, requestMessage.getHeaders().getReplyChannel())
+					.setHeader(MessageHeaders.ERROR_CHANNEL, requestMessage.getHeaders().getErrorChannel())
```

Comment:
I would say we just need to remove these headers and let a super class to populate them!

Yeah, now I see your point about those `ORIGINAL_` headers... Good catch!

Reply:
Hi @artembilan,

When we remove the replyChannel and errorChannel Header mapping, super class is not populating the reply and error channel. This is causing the scatterGather to fail in the nested ScatterGather scenario.

org.springframework.integration.dsl.routers.RouterTests > testNestedScatterGather FAILED
    java.lang.AssertionError at RouterTests.java:545
18 tests completed, 1 failed


Also, we need to keep the headers as the headers populated by Output processor needs to be preserved. This is very common for application to populate the information in headers and later reference them in downstream processing.